src/client/mod.rs

```diff
@@ -513,10 +606,27 @@ impl<C, B> fmt::Debug for Client<C, B> {
 
 /// A `Future` that will resolve to an HTTP Response.
 #[must_use = "futures do nothing unless polled"]
+#[cfg(impl_trait_available)]
+pub struct ResponseFuture<F> {
```

Comment:
Unfortunately this would be a breaking change, since any place someone has before something like `struct Foo(hyper::client::ResponseFuture)` would no longer compile, since they'd need to add the generic in. And with `impl Trait`, they'd have no way of ever define the exact type there, so the generic would need to be propagated to every other type holding one. So in the mean time, *this* case will need to stay with a inner boxed trait object.

(Supposedly it will be possible later to have `struct ResponseFuture { inner: existential type Future<..> }` or something like that, so there's hope!)